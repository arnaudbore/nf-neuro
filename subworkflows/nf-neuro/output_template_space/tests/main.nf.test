nextflow_workflow {

    name "Test Subworkflow OUTPUT_TEMPLATE_SPACE"
    script "../main.nf"
    workflow "OUTPUT_TEMPLATE_SPACE"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfneuro"
    tag "subworkflows/output_template_space"
    tag "subworkflows/registration"
    tag "load_test_data"

    tag "registration/antsapplytransforms"
    tag "registration/tractogram"
    tag "image/applymask"
    tag "betcrop/fslbetcrop"
    tag "utils/templateflow"

    tag "stub"
    options "-stub-run"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "tractometry.zip", "freesurfer.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("Template MNI152NLin2009cAsym - local templates") {
        config "./local.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                input[5] = Channel.empty()
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    ["ch_t1w_tpl", "ch_t2w_tpl"].contains(channel.key)
                                        ? file(subject).name
                                        : [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry
                                            ? file(entry).name
                                            : "" }
                            } ]
                        }
                ).match()},
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["ch_t1w_tpl",
                                        "ch_t2w_tpl",
                                        "ch_registered_anat",
                                        "ch_registered_labels_files",
                                        "ch_registered_nifti_files",
                                        "ch_registered_trk_files",
                                        "mqc",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("Template MNI152NLin2009aAsym - no brain mask") {
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                input[5] = Channel.empty()
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    ["ch_t1w_tpl", "ch_t2w_tpl"].contains(channel.key)
                                        ? file(subject).name
                                        : [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry
                                            ? file(entry).name
                                            : "" }
                            } ]
                        }
                ).match()},
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["ch_t1w_tpl",
                                        "ch_t2w_tpl",
                                        "ch_registered_anat",
                                        "ch_registered_labels_files",
                                        "ch_registered_nifti_files",
                                        "ch_registered_trk_files",
                                        "mqc",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }

    test("Template MNI152NLin2009aAsym - using synthmorph registration") {
        config "./synthmorph.config"
        when {
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        tractometry: it.simpleName == "tractometry"
                        freesurfer: it.simpleName == "freesurfer"
                    }
                input[0] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        file("\${test_data_directory}/mni_masked.nii.gz")
                    ]
                }
                input[1] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.nii.gz")
                        ]
                    ]
                }
                input[2] = Channel.empty()
                input[3] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM_labels_map.nii.gz")
                        ]
                    ]
                }
                input[4] = ch_split_test_data.tractometry.map{
                    test_data_directory -> [
                        [ id: 'test' ], // meta map
                        [
                            file("\${test_data_directory}/IFGWM.trk"),
                            file("\${test_data_directory}/IFGWM_color.trk"),
                            file("\${test_data_directory}/IFGWM_uni.trk")
                        ]
                    ]
                }
                input[5] = ch_split_test_data.freesurfer.map{
                    test_data_directory -> file("\${test_data_directory}/license.txt")
                }
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(
                    workflow.out
                        .findAll{ channel -> !channel.key.isInteger() && channel.value }
                        .collectEntries{ channel ->
                            [(channel.key): ["versions"].contains(channel.key)
                                ? channel.value
                                : channel.value.collect{ subject ->
                                    ["ch_t1w_tpl", "ch_t2w_tpl"].contains(channel.key)
                                        ? file(subject).name
                                        : [ subject[0] ] + subject[1..-1].flatten().collect{ entry -> entry
                                            ? file(entry).name
                                            : "" }
                            } ]
                        }
                ).match()},
                { assert workflow.out
                    .findAll{ channel -> !channel.key.isInteger() }
                    .every{ channel -> ["ch_registered_anat",
                                        "ch_t1w_tpl",
                                        "ch_t2w_tpl",
                                        "ch_registered_labels_files",
                                        "ch_registered_nifti_files",
                                        "ch_registered_trk_files",
                                        "mqc",
                                        "versions"].contains(channel.key)
                        ? channel.value.every{ subject -> subject instanceof ArrayList
                                ? subject.every()
                                : subject }
                        : channel.value.size() == 0 } }
            )
        }
    }
}
