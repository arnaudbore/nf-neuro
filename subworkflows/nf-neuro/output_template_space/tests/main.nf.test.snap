{
    "Template MNI152NLin2009aAsym - using synthmorph registration": {
        "content": [
            {
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test_warped.nii.gz"
                    ]
                ],
                "ch_registered_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_warped.nii.gz"
                    ]
                ],
                "ch_registered_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_warped.nii.gz"
                    ]
                ],
                "ch_registered_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM.trk",
                        "test_IFGWM_color.trk",
                        "test_IFGWM_uni.trk"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_registration_antsapplytransforms_mqc.gif"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_registration_antsapplytransforms_mqc.gif"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                    "versions.yml:md5,4e02408f236932635d1e7585d80d16a1",
                    "versions.yml:md5,5091055c3aacdbd4dccc294cc841b8b3",
                    "versions.yml:md5,513f314aaedec96c49c67ed0bd3bbf4b",
                    "versions.yml:md5,53e0a593f56efd67061357b22888233f",
                    "versions.yml:md5,837bc1b10db6cb6c459b8871725ba254",
                    "versions.yml:md5,9013974da5ab5a7ae91e580948968e75",
                    "versions.yml:md5,957fd24bd8b559411e3a565ac7313f40"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T15:51:42.158522262"
    },
    "Template MNI152NLin2009aAsym - no brain mask": {
        "content": [
            {
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz"
                    ]
                ],
                "ch_registered_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_warped.nii.gz"
                    ]
                ],
                "ch_registered_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_warped.nii.gz"
                    ]
                ],
                "ch_registered_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM.trk",
                        "test_IFGWM_color.trk",
                        "test_IFGWM_uni.trk"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_registration_antsapplytransforms_mqc.gif"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_registration_antsapplytransforms_mqc.gif"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_ants_mqc.gif"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                    "versions.yml:md5,2b72cc481686daf9f4a00587bbf1efa3",
                    "versions.yml:md5,4e02408f236932635d1e7585d80d16a1",
                    "versions.yml:md5,5091055c3aacdbd4dccc294cc841b8b3",
                    "versions.yml:md5,513f314aaedec96c49c67ed0bd3bbf4b",
                    "versions.yml:md5,53e0a593f56efd67061357b22888233f",
                    "versions.yml:md5,957fd24bd8b559411e3a565ac7313f40"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.10.0"
        },
        "timestamp": "2025-10-28T15:50:40.403026792"
    },
    "Template MNI152NLin2009cAsym - local templates": {
        "content": [
            {
                "ch_registered_anat": [
                    [
                        {
                            "id": "test"
                        },
                        "test_t1_warped.nii.gz"
                    ]
                ],
                "ch_registered_labels_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_warped.nii.gz"
                    ]
                ],
                "ch_registered_nifti_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_warped.nii.gz"
                    ]
                ],
                "ch_registered_trk_files": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM.trk",
                        "test_IFGWM_color.trk",
                        "test_IFGWM_uni.trk"
                    ]
                ],
                "ch_t1w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "ch_t2w_tpl": [
                    "template__image_bet.nii.gz"
                ],
                "mqc": [
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_labels_map_registration_antsapplytransforms_mqc.gif"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test_IFGWM_registration_antsapplytransforms_mqc.gif"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "test__registration_ants_mqc.gif"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,08c023826481c9130c3914096cda7f63",
                    "versions.yml:md5,2b72cc481686daf9f4a00587bbf1efa3",
                    "versions.yml:md5,4e02408f236932635d1e7585d80d16a1",
                    "versions.yml:md5,5091055c3aacdbd4dccc294cc841b8b3",
                    "versions.yml:md5,513f314aaedec96c49c67ed0bd3bbf4b",
                    "versions.yml:md5,53e0a593f56efd67061357b22888233f",
                    "versions.yml:md5,957fd24bd8b559411e3a565ac7313f40"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.8"
        },
        "timestamp": "2025-10-24T14:50:02.140523373"
    }
}